File: 61-rotate-list/rotate-list.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || k == 0) return head; // If the list is empty or rotation count is zero, return the list as is
        
        int length = 1; // Length of the linked list
        ListNode *tail = head; // Pointer to the tail of the linked list
        
        // Traverse the list to find the length and the tail node
        while (tail->next) {
            tail = tail->next;
            length++;
        }
        
        // Adjust k to avoid unnecessary rotations
        k = k % length;
        
        // If k is zero, no rotation is needed
        if (k == 0) return head;
        
        // Find the new tail node after rotation
        ListNode *newTail = head;
        for (int i = 1; i < length - k; ++i) {
            newTail = newTail->next;
        }
        
        // Update pointers to perform rotation
        tail->next = head;
        head = newTail->next;
        newTail->next = nullptr;
        
        return head;
    }
};

File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null || k == 0) return head;
        int length = 1;
        ListNode temp = head;
        while(temp.next!=null){
            length++;
            temp = temp.next;
        }
        ListNode tail = temp;
        tail.next = head;
        if(k % length == 0){tail.next = null;return head;}
        int limit = length - (k%length);
        int i = 1;
        while(i < limit+1){
            temp = temp.next;
            i++;
        }
        ListNode newHead = temp.next;
        temp.next = null;
        return newHead;
    }
}